File: solutions/jump-game/jump-game.cpp
// @Title: Jump Game
// @Author: colinjxc
// @Date: 2022-02-15T22:38:08+08:00
// @URL: https://leetcode-cn.com/problems/jump-game


class Solution {
public:
    bool canJump(vector<int>& nums) {
        int far_most = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (far_most < i) 
                return false;
            if (nums[i] + i > far_most) {
                far_most = nums[i] + i;
            }
            if (far_most >= nums.size() - 1)
                return true;
        }
        return far_most >= nums.size() - 1;
    }
};

File: solutions/jump-game/jump-game.py
// @Title: Jump Game
// @Author: colinjxc
// @Date: 2022-01-28T04:01:36+08:00
// @URL: https://leetcode-cn.com/problems/jump-game


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n, far_most = len(nums), 0
        for i in range(n):
            if i <= far_most:
                far_most = max(far_most, i + nums[i])
                if far_most + 1 >= len(nums):
                    return True
            else:
                break
        return False
